### PR TITLE
Mule-9950: Make FreePortFinder thread safe

### PR DESCRIPTION
**Pull request by manuporto**
Changed how Free Port Finder stores the ports being used. Now it's stores them in files (one per used port) and locks them until they are freed.

https://www.mulesoft.org/jira/browse/MULE-9950

### CODE REVIEW

File: core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java [hunk 1]
```diff
@@ -38,24 +46,42 @@ public synchronized Integer find()
...
-            int port = minPortNumber + random.nextInt(portRange);
+            try
+            {
+                int port = minPortNumber + random.nextInt(portRange);
+                String portFile = String.valueOf(port) + LOCK_FILE_EXTENSION;
+                FileChannel channel = FileChannel.open(Paths.get(portFile), CREATE, WRITE);
+                FileLock lock = channel.tryLock();
 
-            if (selectedPorts.contains(port))
+                if (isPortFree(port))
```
Comment (elrodro83) [inline core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java]: should check that the lock was acqired (`lock != null`). Ref: https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#tryLock()

File: core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java [hunk 2]
```diff
@@ -38,24 +46,42 @@ public synchronized Integer find()
     {
         for (int i = 0; i < portRange; i++)
         {
-            int port = minPortNumber + random.nextInt(portRange);
+            try
+            {
+                int port = minPortNumber + random.nextInt(portRange);
+                String portFile = String.valueOf(port) + LOCK_FILE_EXTENSION;
+                FileChannel channel = FileChannel.open(Paths.get(portFile), CREATE, WRITE);
```
Comment (elrodro83) [inline core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java]: use static imports

File: core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java [hunk 3]
```diff
@@ -38,24 +46,42 @@ public synchronized Integer find()
     {
         for (int i = 0; i < portRange; i++)
         {
-            int port = minPortNumber + random.nextInt(portRange);
+            try
+            {
+                int port = minPortNumber + random.nextInt(portRange);
+                String portFile = String.valueOf(port) + LOCK_FILE_EXTENSION;
```
Comment (elrodro83) [inline core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java]: No need to use `String.valueOf`

**Comment by afelisatti** [comment]
@manuporto, please create a separate issue to track this and reference it here.

**Comment by afelisatti** [comment]
Actually, seems this had been reported already: MULE-9950

File: core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java [hunk 4]
```diff
@@ -38,24 +49,49 @@ public synchronized Integer find()
     {
         for (int i = 0; i < portRange; i++)
         {
-            int port = minPortNumber + random.nextInt(portRange);
+            try
+            {
+                int port = minPortNumber + random.nextInt(portRange);
+                String portFile = port + LOCK_FILE_EXTENSION;
```
Comment (elrodro83) [inline core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java]: this line and the previous can be outside of the `try`

File: core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java [hunk 5]
```diff
@@ -71,15 +107,23 @@ public synchronized Integer find()
...
     {
-        if (isPortFree(port))
+        if (isPortFree(port) && locks.containsKey(port) && files.containsKey(port))
         {
-            selectedPorts.remove(port);
+            FileLock lock = locks.get(port);
+            FileChannel file = files.get(port);
+            try {
+                lock.release();
+                file.close();
```
Comment (elrodro83) [inline core/src/test/java/org/mule/tck/junit4/rule/FreePortFinder.java]: remove the objects from the map

**Pull request merged**